src/pybullet_helpers/joint.py

```diff
@@ -0,0 +1,103 @@
+"""Pybullet helper class for joint utilities."""
+from typing import List, NamedTuple, Sequence, Tuple
+
+import numpy as np
+import pybullet as p
+
+from predicators.src.pybullet_helpers.geometry import Pose3D, Quaternion
+
+
+class JointInfo(NamedTuple):
+    """Joint Information from Pybullet."""
+    jointIndex: int
+    jointName: str
+    jointType: int
+    qIndex: int
+    uIndex: int
+    flags: int
+    jointDamping: float
+    jointFriction: float
+    jointLowerLimit: float
+    jointUpperLimit: float
+    jointMaxForce: float
+    jointMaxVelocity: float
+    linkName: str
+    jointAxis: Sequence[float]
+    parentFramePos: Pose3D
+    parentFrameOrn: Quaternion
+    parentIndex: int
+
+    def is_circular(self) -> bool:
+        """Whether the joint is circular or not."""
+        if self.is_fixed():
+            return False
+        return self.jointUpperLimit < self.jointLowerLimit
+
+    def is_fixed(self) -> bool:
+        """Whether the joint is fixed or not."""
+        return self.jointType == p.JOINT_FIXED
+
+
+def get_num_joints(body: int, physics_client_id: int) -> int:
+    """Get the number of joints for a body."""
+    return p.getNumJoints(body, physicsClientId=physics_client_id)
+
+
+def get_joints(body: int, physics_client_id: int) -> List[int]:
+    """Get joint indices for a body."""
+    return list(range(get_num_joints(body, physics_client_id)))
+
+
+def get_joint_info(body: int, joint: int, physics_client_id: int) -> JointInfo:
+    """Get the info for the given joint for a body."""
+    raw_joint_info: List = list(
+        p.getJointInfo(body, joint, physicsClientId=physics_client_id))
+    # Decode the byte strings for joint name and link name
+    raw_joint_info[1] = raw_joint_info[1].decode("UTF-8")
+    raw_joint_info[12] = raw_joint_info[12].decode("UTF-8")
+
+    joint_info = JointInfo(*raw_joint_info)
+    return joint_info
+
+
+def get_joint_infos(body: int, joints: List[int],
```

Comment:
since we've previously used `JointsState` instead of `JointStates`, should we call this `joints_info` instead of `joint_infos`?

Reply:
similarly for the other functions below about joint limits, which makes even more sense to convert because "joint limits" sounds like the limits for a single joint

Reply:
I'm going to look through the codebase for these plural/singular words and get back to you.

Personally I prefer having stuff like JointStates, for some reason JointsState feels weird.

Reply:
I'll advocate for using:

- JointSingular when it refers to a single joint
- JointPlural when it refers to multiple joints

This is the convention at least when describing it in a textbook - e.g. joint positions, joint angles, joint limits

Reply:
okay sounds good to me! i agree, jointsstate is kinda weird

Reply:
Migrated to `JointPositions`

JointState (to match the pybullet definition) should include positions, velocities and forces which we don't use.